### ISSUE TITLE
Pyright doesn't detect coroutine awaits properly.

### ISSUE THREAD

**Issue by kfreezen** [microsoft/pyright#5220]
I've run across a rather odd behavior in pyright's `async/await` detection.

Here's a test case
```
class TestCase:
    async def test_fn(self, r):
        return {}

    async def case1(self):
        b = []
        m = await self.test_fn(b)
        m["test"]

    async def case2(self):
        m = await self.test_fn([])
        m["test"]

    async def case3(self):
        m = await self.test_fn(None)
        m["test"]
```

For some reason, in `case2`, `await test_fn` is reported as returning a Coroutine, when it should be returning a `dict[Unknown, Unknown]`

`case1` and `case3` seem to detect test_fn's return type correctly.

The pyright output for this:
```
/path/test-pyright/test.py
  /path/test-pyright/test.py:12:9 - error: "__getitem__" method not defined on type "Coroutine[Any, Any, dict[Unknown, Unknown]]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Expected output would be
```
0 errors, 0 warnings, 0 informations
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

You can work around it in the meantime by providing a type annotation for the `r` parameter or the return type for method `test_fn`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.
